lib/SILOptimizer/Utils/InstOptUtils.cpp

```diff
@@ -472,8 +472,8 @@ SILValue swift::castValueToABICompatibleType(SILBuilder *builder,
   if (srcTy == destTy)
     return value;
 
-  assert(srcTy.isAddress() == destTy.isAddress()
-         && "Addresses aren't compatible with values");
+  if (srcTy.isAddress() == destTy.isAddress())
+    return nullptr;
```

Comment:
Is this change intentional?

Reply:
Yes. This assertion was triggered after the above change. I suspect some other optimization caused the store instruction to trigger this assertion. This was the simplest fix but, maybe I should try to do some more digging.

Reply:
It's not part of this APIs contract to return nullptr and it doesn't seem to be anticipated by its callers.

Reply:
I am now thinking that this assertion shouldn't exist. The next condition `srcTy.isAddress() && destTy.isAddress()` can never succeed if this assertion is triggered (if `bool != bool` then `bool && bool == false`).

Reply:
This is valid code
`assert(A == B); if (A && B)`

Reply:
Yes, you're right. I got confused for a second :P 

I'm not sure what the best solution is, but it seems to be OK when I remove the assertion (not a reason on its own to remove the assertion, though).